src/AzureAuth/Commands/Ado/CommandPat.cs

```diff
@@ -63,16 +98,104 @@ private enum OutputMode
+            var manager = new PatManager(cache, client);
+
+            using (new CrossPlatLock(LockfilePath))
+            {
+                var pat = manager.GetPatAsync(this.PatOptions()).Result;
+
+                // Do not use logger to avoid printing PATs into log files.
+                Console.WriteLine(FormatPat(pat, this.Output));
+            }
+
             return 0;
         }
+
+        private static string FormatPat(PatToken pat, OutputMode output) => output switch
+        {
+            OutputMode.None => string.Empty,
+            OutputMode.Status => $"\"{pat.DisplayName}\" valid until {pat.ValidTo:O}",
+            OutputMode.Token => pat.Token,
+            OutputMode.Base64 => pat.Token.Base64(),
+            OutputMode.Header => pat.Token.AsHeader(AzureAuth.Ado.Authorization.Basic),
+            OutputMode.HeaderValue => pat.Token.AsHeaderValue(AzureAuth.Ado.Authorization.Basic),
+            OutputMode.Json => pat.AsJson(),
+            _ => throw new ArgumentOutOfRangeException(nameof(output)),
+        };
+
+        private TokenResult AccessToken(IPublicClientAuth publicClientAuth, CommandExecuteEventData eventData)
+        {
+            return publicClientAuth.Token(
+                AzureAuth.Ado.Constants.AdoParams,
+                AccessTokenAuthModes,
+                this.Domain,
+                AccessTokenPrompt,
+                AccessTokenTimeout,
+                eventData);
+        }
+
+        private PatOptions PatOptions()
+        {
+            return new PatOptions
+            {
+                Organization = this.Organization,
+                DisplayName = this.DisplayName,
+                Scopes = this.Scopes,
+            };
+        }
+
+        private IPatClient Client(string accessToken)
+        {
+            var baseUrl = new Uri(string.Join('/', AzureAuth.Ado.Constants.BaseUrl, this.Organization));
+            var credentials = new VssOAuthAccessTokenCredential(accessToken);
+            var connection = new VssConnection(baseUrl, credentials);
+            var tokensHttpClientWrapper = new TokensHttpClientWrapper(connection);
+            return new PatClient(tokensHttpClientWrapper);
+        }
+
+        private IPatCache Cache()
+        {
+            var storageProperties = new StorageCreationPropertiesBuilder(
+                PatStorageParameters.CacheFileName,
+                CacheDirectory)
+            .WithMacKeyChain(
+                PatStorageParameters.MacOSServiceName,
+                PatStorageParameters.MacOSAccountName)
+            .WithLinuxKeyring(
+                PatStorageParameters.LinuxKeyRingSchemaName,
+                PatStorageParameters.LinuxKeyRingCollection,
+                PatStorageParameters.LinuxKeyRingLabel,
+                PatStorageParameters.LinuxKeyRingAttr1,
+                PatStorageParameters.LinuxKeyRingAttr2)
+            .Build();
+
+            // TODO: We probably need to do a `storage.VerifyPersistence` check

```

Comment:
In my experience, the cache persistency check has only failed on mac, when there is some unexplainable issue with the KeyChain. And in that case, I'm still torn on whether we should fail all up, or proceed with effectively no cache, which could result in an auth prompt and new PAT on every run.

Reply:
If this were being called in parallel with the expectation that the user would get a single PAT from caching the first call I think we'd end up with a bad user experience if we ignored caching failures. It could lead to a "PAT explosion" where you wanted 1 PAT, but now you have 100 because none of them got cached between calls.